src/machine/eval.rs

```diff
@@ -0,0 +1,157 @@
+        &mut self,
+        mut t1: Box<Term>,
+        t2: Box<Term>,
+        t3: Box<Term>,
+    ) -> (bool, Term) {
+        // If t1 is a literal, we should be able to evaluate the conditional
+        if let box Term::Lit(lit) = t1 {
+            match lit {
+                // If t1 is true, evaluate to t2.
+                Literal::Bool(true) => (true, *t2),
+                // If t1 is false, evaluate to t3.
+                Literal::Bool(false) => (true, *t3),
+                // If t1 is any other literal, panic
+                lit => panic!("Found non-boolean literal {} in condition", lit),
+            }
+        // If t1 is not a literal, evaluate it in place and return (if t1 then t2 else t3)
+        } else {
+            (self.step_in_place(t1.as_mut()), Term::Cond(t1, t2, t3))
+        }
+    }
+
+    /// Evaluation step for binary operations (t1 op t2)
+    #[inline(always)]
+    fn step_bin_op(&mut self, op: BinOp, t1: Box<Term>, t2: Box<Term>) -> (bool, Term) {
+        use BinOp::*;
+        use Literal::*;
+
+        match (op, t1, t2) {
+            (op, box Lit(l1), box Lit(l2)) => (true, Lit(native_bin_op(op, l1, l2))),
+            // If op is && and t1 is false evaluate to false
+            (And, box Lit(Bool(false)), _) => (true, Lit(Bool(false))),
+            // If op is || and t1 is true evaluate to true
+            (Or, box Lit(Bool(true)), _) => (true, Lit(Bool(true))),
+            // If t2 is not a literal, evaluate it.
+            (op, t1 @ box Lit(_), mut t2) => {
+                (self.step_in_place(t2.as_mut()), Term::BinaryOp(op, t1, t2))
+            }
+            // If t1 is not a literal, evaluate it.
+            (op, mut t1, t2) => (self.step_in_place(t1.as_mut()), Term::BinaryOp(op, t1, t2)),
+        }
+    }
+
+    /// Evaluation step for unary operations (op t1)
+    #[inline(always)]
+    fn step_un_op(&mut self, op: UnOp, mut t1: Box<Term>) -> (bool, Term) {
+        // If t1 is a literal, do the operation.
+        if let box Term::Lit(lit) = t1 {
+            (true, Term::Lit(native_un_op(op, lit)))
+        // If t1 is not a literal, evaluate it.
+        } else {
+            (self.step_in_place(t1.as_mut()), Term::UnaryOp(op, t1))
+        }
+    }
+
+    /// Evaluation step for the fixed-point operation (fix t1)
+    fn step_fix(&mut self, mut t1: Box<Term>) -> (bool, Term) {
+        // If t1 is an abstraction (\. t2), replace the argument of t1 by (fix t1) inside t2
+        // and evaluate to t2.
+        if let box Term::Abs(box ref t2) = t1 {
+            let mut t2 = t2.clone();
+            t2.replace(0, &mut Term::Fix(t1));
+            (true, t2)
+        // If t1 is not an abstraction, evaluate it.
+        } else {
+            (self.step_in_place(t1.as_mut()), Term::Fix(t1))
+        }
+    }
+}
+
+/// Evaluation step for beta reduction ((λ. body) arg)
+#[inline(always)]
+fn step_beta_reduction(mut body: Box<Term>, mut arg: Box<Term>) -> (bool, Term) {
```

Comment:
I think you can change this one to be a method of `Machine` too for the sake of consistency.